File: adapters/rust/src/emit.rs
use crate::contract::JsonMap;
use crate::model::Context;
use anyhow::{Context as AnyhowContext, Result};
use serde_json::Value;

pub(crate) fn render(context: &Context, repository: &str) -> Result<String> {
    let mut header = JsonMap::new();
    header.insert("adapter_version".into(), Value::String("0.1.0".into()));
    header.insert("language".into(), Value::String("rust".into()));
    header.insert("record".into(), Value::String("lexicon".into()));
    header.insert("repository".into(), Value::String(repository.into()));
    header.insert("schema_version".into(), Value::Number(1.into()));
    let mut values = vec![Value::Object(header)];
    let mut facts: Vec<Value> = context
        .facts
        .nodes
        .values()
        .chain(context.facts.edges.values())
        .chain(context.facts.unresolved.values())
        .cloned()
        .collect();
    facts.sort_by_key(fact_sort_key);
    values.extend(facts);
    values
        .into_iter()
        .map(|value| serde_json::to_string(&value).context("cannot serialize fact"))
        .collect::<Result<Vec<_>>>()
        .map(|lines| format!("{}\n", lines.join("\n")))
}

pub(crate) fn fact_sort_key(value: &Value) -> (u8, String, String, String, String, Vec<String>) {
    let object = value.as_object().expect("fact records are objects");
    match object.get("record").and_then(Value::as_str).unwrap_or("") {
        "node" => (
            0,
            field(object, "id"),
            field(object, "kind"),
            field(object, "path"),
            field(object, "qualified_name"),
            Vec::new(),
        ),
        "edge" => (
            1,
            field(object, "source"),
            field(object, "target"),
            field(object, "relation"),
            String::new(),
            span_sort_key(object.get("span")),
        ),
        _ => (
            2,
            field(object, "source"),
            field(object, "relation"),
            field(object, "expression"),
            field(object, "reason"),
            span_sort_key(object.get("span")),
        ),
    }
}

fn field(object: &JsonMap, name: &str) -> String {
    object
        .get(name)
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string()
}

fn span_sort_key(value: Option<&Value>) -> Vec<String> {
    let Some(span) = value.and_then(Value::as_object) else {
        return vec![String::new(); 5];
    };
    [
        "path",
        "start_line",
        "start_column",
        "end_line",
        "end_column",
    ]
    .iter()
    .map(|key| match span.get(*key) {
        Some(Value::String(value)) => value.clone(),
        Some(value) => value.to_string(),
        None => String::new(),
    })
    .collect()
}

File: lexicon/internal/interstack/emit.go
package interstack

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

func Encode(result Result) ([]byte, error) {
	repository := result.Repository
	if repository == "" {
		repository = "repository"
	}
	header := map[string]any{
		"adapter_version": AdapterVersion,
		"language":        Language,
		"mode":            "full",
		"record":          "lexicon",
		"repository":      repository,
		"schema_version":  1,
	}
	var output bytes.Buffer
	if err := writeJSONLine(&output, header); err != nil {
		return nil, err
	}
	nodes := append([]Node(nil), result.Nodes...)
	sort.Slice(nodes, func(left, right int) bool {
		if nodes[left].ID != nodes[right].ID {
			return nodes[left].ID < nodes[right].ID
		}
		if nodes[left].Kind != nodes[right].Kind {
			return nodes[left].Kind < nodes[right].Kind
		}
		return nodes[left].QualifiedName < nodes[right].QualifiedName
	})
	for _, node := range nodes {
		record := map[string]any{
			"id":             node.ID,
			"kind":           node.Kind,
			"name":           node.Name,
			"path":           node.Path,
			"qualified_name": node.QualifiedName,
			"record":         "node",
		}
		if len(node.Attributes) > 0 {
			record["attributes"] = node.Attributes
		}
		if node.Span != nil {
			record["span"] = node.Span
		}
		if err := writeJSONLine(&output, record); err != nil {
			return nil, err
		}
	}
	edges := append([]factEdge(nil), result.Edges...)
	sort.Slice(edges, func(left, right int) bool {
		return edgeSortKey(edges[left]) < edgeSortKey(edges[right])
	})
	for _, edge := range edges {
		record := map[string]any{
			"record":   "edge",
			"relation": edge.Relation,
			"source":   edge.Source,
			"target":   edge.Target,
		}
		if edge.Span != nil {
			record["span"] = edge.Span
		}
		if len(edge.Attributes) > 0 {
			record["attributes"] = edge.Attributes
		}
		if err := writeJSONLine(&output, record); err != nil {
			return nil, err
		}
	}
	unresolved := append([]factUnresolved(nil), result.Unresolved...)
	sort.Slice(unresolved, func(left, right int) bool {
		return unresolvedSortKey(unresolved[left]) < unresolvedSortKey(unresolved[right])
	})
	for _, item := range unresolved {
		record := map[string]any{
			"expression": item.Expression,
			"reason":     item.Reason,
			"record":     "unresolved",
			"relation":   item.Relation,
			"source":     item.Source,
		}
		if item.Span != nil {
			record["span"] = item.Span
		}
		if len(item.Attributes) > 0 {
			record["attributes"] = item.Attributes
		}
		if err := writeJSONLine(&output, record); err != nil {
			return nil, err
		}
	}
	return output.Bytes(), nil
}

func writeJSONLine(output *bytes.Buffer, value any) error {
	data, err := json.Marshal(value)
	if err != nil {
		return err
	}
	output.Write(data)
	output.WriteByte('\n')
	return nil
}

func stableNodeID(kind, identity string) string {
	payload := "lexicon:v1\x00" + Language + "\x00" + kind + "\x00" + identity
	digest := sha256.Sum256([]byte(payload))
	return "sha256:" + hex.EncodeToString(digest[:])
}

func edgeSortKey(edge factEdge) string {
	return strings.Join([]string{edge.Source, edge.Target, edge.Relation, spanSortKey(edge.Span)}, "\x00")
}

func unresolvedSortKey(item factUnresolved) string {
	return strings.Join([]string{item.Source, item.Relation, item.Expression, item.Reason, spanSortKey(item.Span)}, "\x00")
}

func spanSortKey(span *Span) string {
	if span == nil {
		return ""
	}
	return fmt.Sprintf("%s:%010d:%010d:%010d:%010d", span.Path, span.StartLine, span.StartColumn, span.EndLine, span.EndColumn)
}
